service/storage_service.cc

```diff
@@ -969,10 +969,12 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     bool is_member = tmptr->is_member(endpoint);
+    bool do_notify_joined = !is_member && !endpoints_to_remove.contains(endpoint);
+
     // Update pending ranges after update of normal tokens immediately to avoid
     // a race where natural endpoint was updated to contain node A, but A was
     // not yet removed from pending endpoints
-    if (!is_member) {
+    if (do_notify_joined) {
```

Comment:
This place worries me. The thing is that subsequent call to `update_normal_tokens` will assert if the endpoint will not be met in the topology. However, this `if (do_notify_joined)` is not necessarily in sync with this requirement.

Reply:
I'll see if in all cases that the endpoint is in endpoints_to_remove its owned_tokens is empty.
If this is the case, we can skip both update_topology and update_normal_tokens (that returns immediately `if (tokens.empty())`)